lib/mutgen.c

```diff
@@ -433,28 +710,289 @@ mutgen_generate(mutgen_t *self, tsk_table_collection_t *tables, int flags)
+            }
+        }
+        if (keep_mutation) {
+            bottom_mutation[mut->node] = mut;
+            j++;
+        }
+    }
+    /* Reset the mapping for the next site */
+    for (mut = site->mutations; mut != NULL; mut = mut->next) {
+        bottom_mutation[mut->node] = NULL;
+    }
+
+out:
+    return ret;
+}
+
+static int MSP_WARN_UNUSED
+mutgen_apply_mutations(mutgen_t *self, tsk_table_collection_t *tables)
+{
+    int ret = 0;
+    const tsk_id_t *I, *O;
+    const tsk_edge_table_t edges = tables->edges;
+    const tsk_node_table_t nodes = tables->nodes;
+    const tsk_id_t M = (tsk_id_t) edges.num_rows;
+    tsk_id_t tj, tk, mutation_id;
+    tsk_id_t *parent = NULL;
+    mutation_t **bottom_mutation = NULL;
+    double left, right;
+    avl_node_t *avl_node;
+    site_t *site;
+    size_t j;
+
+    parent = malloc(nodes.num_rows * sizeof(*parent));
+    bottom_mutation = malloc(nodes.num_rows * sizeof(*bottom_mutation));
+    if (parent == NULL || bottom_mutation == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(parent, 0xff, nodes.num_rows * sizeof(*parent));
+    for (j = 0; j < nodes.num_rows; j++) {
+        bottom_mutation[j] = NULL;
+    }
+
+    if (!tsk_table_collection_has_index(tables, 0)) {
+        ret = tsk_table_collection_build_index(tables, 0);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+    I = tables->indexes.edge_insertion_order;
+    O = tables->indexes.edge_removal_order;
+    tj = 0;
+    tk = 0;
+    mutation_id = 0;
+    left = 0;
+    avl_node = self->sites.head;
+    while (tj < M || left < tables->sequence_length) {
+        while (tk < M && edges.right[O[tk]] == left) {
+            parent[edges.child[O[tk]]] = TSK_NULL;
+            tk++;
+        }
+        while (tj < M && edges.left[I[tj]] == left) {
+            parent[edges.child[I[tj]]] = edges.parent[I[tj]];
+            tj++;
+        }
+        right = tables->sequence_length;
+        if (tj < M) {
+            right = TSK_MIN(right, edges.left[I[tj]]);
+        }
+        if (tk < M) {
+            right = TSK_MIN(right, edges.right[O[tk]]);
+        }
+
+        /* Tree is now ready. We look at each site on this tree in turn */
+        while (avl_node != NULL) {
+            site = (site_t *) avl_node->item;
+            if (site->position >= right) {
+                break;
+            }
+            sort_mutations(site);
+            ret = mutgen_choose_alleles(self, parent, bottom_mutation,
+                    nodes.num_rows, site, mutation_id);
+            if (ret < 0) {
+                goto out;
+            }
+            // TODO: slightly worried about this
+            mutation_id += (int) site->mutations_length;
+            if (site->new && site->mutations_length == 0) {
+                avl_delete_node(&self->sites, avl_node);
+            }
+
+            avl_node = avl_node->next;
```

Comment:
You need to get the next **before** deleting the avl node. Something like
``
current = avl_node;
avl_node = avl_node->next;
if (site->new && site->mutations_length == 0) {
￼        avl_delete_node(&self->sites, avl_node);
￼}
```
Also, is ``avl_delete_node`` what we use elsewhere? Some of the AVL functions malloc and free memory. We do **not** want these.

Reply:
Ah, I see. Elsewhere we use `avl_unlink_node`, but since our [avl tree's freeitem is NULL](https://github.com/tskit-dev/msprime/pull/946/files#diff-2855ce2e7aad0c9d8740b3c94e165a5aR399) I think they're the same:
```
/* Deletes a node from the tree. Returns immediately if the node is NULL.
 * If the tree's freeitem is not NULL, it is invoked on the item.
 * If it is, returns the item.
 * O(lg n) */
extern void *avl_delete_node(avl_tree_t *, avl_node_t *);
```
I'll change it anyhow.

Reply:
I'm not sure it is the same, I have a feeling it also frees the AVL node itself (not just the item). We know unlink_node works, so let's stick with it.